### DISCUSSION TITLE
Quay 3.8 Docker v1 deprecation

### DISCUSSION THREAD

**Discussion by thomoe** [quay/quay#3121] [Q&A]
Hi,

I just asked the same in quay-sig in Google Groups, after seeing a couple of links from there to quay/discussions here, I thought here might be the better place to ask.

I'm planning to upgrade a quay instance from 3.7 to 3.8. (It's an intermediate step, further upgrades are planned soon.) The Red Hat release notes for 3.8 mention

"Support for Docker v1 is now deprecated and will be removed in a future release of Red Hat Quay. Users must now opt-in to enable Docker v1 support. Users should migrate any stored images in Docker v1 format to the OCI image format to avoid potential loss of data."

I have following questions:

1. How do I opt in to Docker v1 support? I didn't see a new config field for this purpose, nor an option in the Web UI.
2. Is there a way to find all existing Docker v1 images in quay?

While investigating, I connected to the backend database to have a look under the hood. So far, I found the mediatype and manifest table, and I queried for all manifests with mediatype 13/14  (application/vnd.docker.distribution.manifest.v1+*).
Is the manifest version equivalent to the image version?

I did pull an older docker image with digest sha256:c28xxx from quay 3.7 with mediatype "application/vnd.docker.distribution.manifest.v1+json", re-tagged it and pushed it to another quay 3.7 instance. Checking the manifest table in that backend-db there, I noticed

* the media_type_id changed to 16 (application/vnd.docker.distribution.manifest.v2+json)
* the digest changed to sha256:f58xxx

I re-tagged the same image and pushed it to another quay 3.8 instance, and in the manifest table,

* the media_type_id changed to 16 (application/vnd.docker.distribution.manifest.v2+json)
* the digest changed to sha256:f58xxx (same as pushing to other 3.7 quay)

I'd appreciate any support. I did try to find information by myself first, in release documentation, configuration manual and general google search. Unfortunately, so far without success.

**Comment by ibazulic** [comment]
Hey @thomoe ,

I did reply to you on quay-sig, but I'll reply here as well. The deprecation notice for Docker `v1` concerns Docker `v1` registry protocol, which is not in use for years. So unless you're using Docker `<1.7` or even older versions, then chances you have a `v1` image stored are non-existent. Docker `v2` schema 1 is still fully supported on Quay, even though Docker decided to deprecate both pushing and pulling of those images (`skopeo` and `podman` both still understand and support the schema fully).

Note also that Quay 3.8 is very old and is not officially supported anymore. So I'd very much recommend you move to 3.12 as soon as possible.

HTH!

**Reply by thomoe** [reply]
Hi @ibazulic,

First of all, thanks for the quick reply. It helps already with my understanding, but I still have two questions:

Our quay installation is in use for quite some time, we have images pushed more than 8 years ago which might very well have been pushed with docker < 1.7. I'm reasonably sure everyone is using newer docker versions by now. However, losing access to existing images wouldn't be acceptable, I'm afraid. (My concern is based on the comment in the [Red Hat Quay Feature Tracker](https://docs.redhat.com/en/documentation/red_hat_quay/3.8/html/red_hat_quay_release_notes/rn-3-800#quay-feature-tracker): "*Support for Docker v1 is now deprecated and will be removed in a future release of Red Hat Quay. Users must now opt-in to enable Docker v1 support. **Users should migrate any stored images in Docker v1 format to the OCI image format to avoid potential loss of data.*** ")

The Release documentation says "_Users must now opt-in to enable Docker v1 support._". I wasn't able to find how this opt in is supposed to work. I'd be much more confident with our upgrade if I knew how to enable it in case we see any issues.

Is this about the fields `FEATURE_RESTRICTED_V1_PUSH` and `V1_PUSH_WHITELIST` moving from "Miscellaneous" to "Legacy" in the documentation? Neither field is set in our current configuration, if that helps.

Also, is there a way to get a list of all (if any) affected images? Either via Quay UI or by querying the backend DB directly, both would be fine for me. 

Thanks again, and best regards

**Reply by ibazulic** [reply]
@thomoe ,

Thanks for your reply! The two fields would only impact pushes of new content, for instance if you would **today** use an ancient Docker client, to the registry. I believe that we incorporated a backfill worker that would, during database migration, run and convert `v1` images to their proper format, before all tables are removed. This might have happened during one of the previous migrations too, at this point I'm not exactly sure but I'll try to find out. The two flags were moved to `legacy` status because we did in fact remove all references to the old model which means removal of all tables connected to it as well. The flags themselves are disabled unless explicitly set in Quay's `config.yaml` file which is why you're not seeing them.
